solidity/src/processor/libs/ProcessorMessage.sol

```diff
@@ -0,0 +1,367 @@
+        }
+
+        bool hasRetryLogic = message[offset] != 0;
+        offset += 1;
+
+        RetryLogic memory retryLogic;
+        if (hasRetryLogic) {
+            (retryLogic, offset) = parseRetryLogic(message, offset);
+        }
+
+        subroutine.atomic.functions = functions;
+        subroutine.atomic.hasRetryLogic = hasRetryLogic;
+        subroutine.atomic.retryLogic = retryLogic;
+
+        return (subroutine, offset);
+    }
+
+    /// @notice Parse non-atomic subroutine from bytes
+    /// @param message Raw message bytes
+    /// @param startOffset Starting position in byte array
+    /// @return subroutine Parsed non-atomic subroutine
+    /// @return newOffset Next position in byte array
+    function parseNonAtomicSubroutine(bytes calldata message, uint256 startOffset)
+        internal
+        pure
+        returns (Subroutine memory subroutine, uint256 newOffset)
+    {
+        uint256 offset = startOffset;
+
+        uint64 functionCount = uint64(bytes8(message[offset:offset + 8]));
+        offset += 8;
+
+        NonAtomicFunction[] memory functions = new NonAtomicFunction[](functionCount);
+        for (uint64 i = 0; i < functionCount; i++) {
+            (functions[i], offset) = parseNonAtomicFunction(message, offset);
+        }
+
+        subroutine.nonAtomic.functions = functions;
+
+        return (subroutine, offset);
+    }
+
+    /// @notice Parse atomic function from bytes
+    /// @param message Raw message bytes
+    /// @param startOffset Starting position in byte array
+    /// @return func Parsed atomic function
+    /// @return newOffset Next position in byte array
+    function parseAtomicFunction(bytes calldata message, uint256 startOffset)
+        internal
+        pure
+        returns (AtomicFunction memory func, uint256 newOffset)
+    {
+        uint256 offset = startOffset;
+
+        // Parse 20 bytes for address
+        address contractAddress = address(bytes20(message[offset:offset + 20]));
+        offset += 20;
+
+        func.contractAddress = contractAddress;
+        return (func, offset);
+    }
+
+    /// @notice Parse non-atomic function from bytes
+    /// @param message Raw message bytes
+    /// @param startOffset Starting position in byte array
+    /// @return func Parsed non-atomic function
+    /// @return newOffset Next position in byte array
+    function parseNonAtomicFunction(bytes calldata message, uint256 startOffset)
+        internal
+        pure
+        returns (NonAtomicFunction memory func, uint256 newOffset)
+    {
+        uint256 offset = startOffset;
+
+        // Parse 20 bytes for address
+        address contractAddress = address(bytes20(message[offset:offset + 20]));
+        offset += 20;
+        func.contractAddress = contractAddress;
+
+        func.hasRetryLogic = message[offset] != 0;
+        offset += 1;
+
+        if (func.hasRetryLogic) {
+            (func.retryLogic, offset) = parseRetryLogic(message, offset);
```

Comment:
maybe we can add some comments here about how `parseRetryLogic` increments the offset? was a little confused before looking there

Reply:
Going to rework it, not worth adding comments for now as it wont be present in next PR